### PR TITLE
feat(rest): Add rest features to sw360 portal

### PR DESCRIPTION
**Pull request by maierthomas**
This pull request will add the following rest features to sw360:
- authorization-server
- resource-server (without HAL-Browser)

The gradle build files were converted to maven pom files.

### CODE REVIEW

**Comment by mcjaeger** [comment]
WOW PR triple six!!!

**Comment by maxhbr** [comment]
Maven logs the following error on my machine, maybe move the variable to the top level pom
```
duplicate declaration of version ${springframework.boot.version} @ org.eclipse.sw360:authorization-server:[unknown-version], bsinno-sw360/rest/authorization-server/pom.xml, line 47, column 21
duplicate declaration of version ${springframework.boot.version} @ org.eclipse.sw360:resource-server:[unknown-version], bsinno-sw360/rest/resource-server/pom.xml, line 47, column 21
```

**Comment by maxhbr** [comment]
This PR moves SW360 from 145 dependencies to 223 distinct dependencies.

**Comment by maxhbr** [comment]
Will it still be possible to build fat-jars as an option?

**Comment by maierthomas** [comment]
Removed duplicate declarations.

All dependencies status
Dependencies (current Master): 144
Dependencies (this Branch): 211

Multiple dependencies status
current master:

> 2 antlr:antlr:jar
> 2 com.jcraft:jsch:jar
> 3 commons-logging:commons-logging:jar
> 2 dom4j:dom4j:jar
> 2 org.apache.httpcomponents:httpclient:jar
> 2 org.apache.poi:poi:jar
> 2 org.apache.poi:poi-ooxml:jar
> 2 org.apache.poi:poi-ooxml-schemas:jar
> 2 org.apache.xmlbeans:xmlbeans:jar

this branch:

> 2 antlr:antlr:jar
> 2 com.fasterxml.jackson.core:jackson-annotations:jar
> 2 com.fasterxml.jackson.core:jackson-core:jar
> 2 com.fasterxml.jackson.core:jackson-databind:jar
> 2 com.jcraft:jsch:jar
> 3 commons-logging:commons-logging:jar
> 2 dom4j:dom4j:jar
> 2 joda-time:joda-time:jar
> 3 org.apache.httpcomponents:httpclient:jar
> 2 org.apache.httpcomponents:httpcore:jar
> 2 org.apache.poi:poi:jar
> 2 org.apache.poi:poi-ooxml:jar
> 2 org.apache.poi:poi-ooxml-schemas:jar
> 2 org.apache.xmlbeans:xmlbeans:jar
> 2 org.slf4j:jcl-over-slf4j:jar

**Comment by maierthomas** [comment]
@adrian-evo  can you please have a look.
Can you please verify this branch and check if it is stable?
Thank you.

**Comment by adrian-evo** [comment]
Tested and the branch is stable. No issues found.

File: rest/pom.xml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        <groupId>org.eclipse.sw360</groupId>
+        <artifactId>sw360</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>rest</artifactId>
+    <packaging>pom</packaging>
+
+    <properties>
+        <springframework.boot.version>1.5.8.RELEASE</springframework.boot.version>
```
Comment (maxhbr) [inline rest/pom.xml]: It might be more consistent to move these properties to the top level pom.

**Comment by maxhbr** [comment]
for testing it might be useful to try a full roundtrip with https://github.com/stevinus1/sw360rest_tools

**Comment by maxhbr** [comment]
The deployed wars only return 404, that does not yet look like it is working at all (or am I calling the wrong URL):

Same result when trying to get the auth token via the gradle script and `AccessTokenPrinter`

**Comment by maxhbr** [comment]
Instead of `http://localhost:8080/resource-server-3.1.0-SNAPSHOT/api/` it would be better if the url were something like `http://localhost:8080/rest/api/`or even `http://localhost:8080/api/`

File: rest/pom.xml (hunk 1 above)
Comment (maierthomas) [reply rest/pom.xml]: switched properties to top level pom

**Comment by maierthomas** [comment]
_Rest update_

**Current endpoints:**
Actually the rest API uses the following application context paths:

(Authorization-Server)
http://localhost:8080/authorization/oauth/authorize
http://localhost:8080/authorization/oauth/check_token
http://localhost:8080/authorization/oauth/token

(Resource-Server)
http://localhost:8080/resource/api/authorize
http://localhost:8080/resource/api/projects
http://localhost:8080/resource/api/components
....


**Generate Access Token:**
You can generate the oauth2 access token by calling the following endpoint address.
It is necessary to use the HTTP basic authentication and the Liferay credentials for the grant type:

```
e.g. (http):
curl -X POST --user '<CLIENT-ID>:<CLIENT-SECRET>' -d 'grant_type=password&username=<LIFERAY-USER-LOGIN>&password=<LIFERAY-USER-PASSWORD>' http://locahost:8080/authorization/oauth/token
   Response: {"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsic3czNjAtUkVTVC1BUEkiXSwidXNlcl9uYW1lIjoiYWRtaW5Ac3czNjAub3JnIiwic2NvcGUiOlsic3czNjAucmVhZCIsInN3MzYwLndyaXRlIl0sImV4cCI6MTUxMTgxNDUyNSwiYXV0aG9yaXRpZXMiOlsiUk9MRV9TVzM2MF9VU0VSIl0sImp0aSI6IjcyYWIxOGRlLWQ1NmQtNDM2OS1iNmFlLWEwYWViZDE1MDg0NiIsImNsaWVudF9pZCI6InRydXN0ZWQtc3czNjAtY2xpZW50In0.e3biYixvg8UZrLVxP4ts_jNemyaCkdVq9s0d1NCExrSN_UaX0wX085RUJqLL43aRi8ZelsgUjSqp0eR9ev5jObFZ5nRqRRwnTBpi9RmGfOxxav9Vjq8DOLAf7sQzIR2CFFtBIcC-WLGwKYcQoUrk6zyux7BxgdxdLp4WNx_ERz0Nx1uazrz4UieIRKXZ474AKAXbx9Q5gwC-uYSx6HHcePHV7TTeFCQYEN7zG9yCJsabr6nZSaflNVw8KE2Ot0nMiUF6QDRST4Nh6-aLFAHkV59dKOXjTdeZC760RpdSXbm_0XLL9aFTtdgEMiZc2B2iSFjY-XcMJJqmYlesLKVMBw","token_type":"bearer","expires_in":3599,"scope":"sw360.read sw360.write","jti":"72ab18de-d56d-4369-b6ae-a0aebd150846"}
```
```
e.g. (https):
curl -X POST --user '<CLIENT-ID>:<CLIENT-SECRET>' -d 'grant_type=password&username=<LIFERAY-USER-LOGIN>&password=<LIFERAY-USER-PASSWORD>' http://locahost:8080/authorization/oauth/token
   Response: {"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsic3czNjAtUkVTVC1BUEkiXSwidXNlcl9uYW1lIjoiYWRtaW5Ac3czNjAub3JnIiwic2NvcGUiOlsic3czNjAucmVhZCIsInN3MzYwLndyaXRlIl0sImV4cCI6MTUxMTgxNDUyNSwiYXV0aG9yaXRpZXMiOlsiUk9MRV9TVzM2MF9VU0VSIl0sImp0aSI6IjcyYWIxOGRlLWQ1NmQtNDM2OS1iNmFlLWEwYWViZDE1MDg0NiIsImNsaWVudF9pZCI6InRydXN0ZWQtc3czNjAtY2xpZW50In0.e3biYixvg8UZrLVxP4ts_jNemyaCkdVq9s0d1NCExrSN_UaX0wX085RUJqLL43aRi8ZelsgUjSqp0eR9ev5jObFZ5nRqRRwnTBpi9RmGfOxxav9Vjq8DOLAf7sQzIR2CFFtBIcC-WLGwKYcQoUrk6zyux7BxgdxdLp4WNx_ERz0Nx1uazrz4UieIRKXZ474AKAXbx9Q5gwC-uYSx6HHcePHV7TTeFCQYEN7zG9yCJsabr6nZSaflNVw8KE2Ot0nMiUF6QDRST4Nh6-aLFAHkV59dKOXjTdeZC760RpdSXbm_0XLL9aFTtdgEMiZc2B2iSFjY-XcMJJqmYlesLKVMBw","token_type":"bearer","expires_in":3599,"scope":"sw360.read sw360.write","jti":"72ab18de-d56d-4369-b6ae-a0aebd150846"}
```

**Rest addon/plugin:**
The HAL-Browser is not a part of this pull request.
But beside the HAL-Browser you can use several plugins/addons for Firefox or chrome.
I recommend the postman addon (chrome).
In postman you can directly authorize with oauth 2.0 to generate an access token.

postman1

**Authorization settings:**

```
Auth URL: http://192.168.1.200:8080/authorization/oauth/auhtorize
Access Token: http://192.168.1.200:8080/authorization/oauth/token?grant_type=password&username=IFERAY-USER>&password=<LIFERAY-USER-PW>
ClientID: x
ClientSecret: x
Scope (optional): sw360.read sw360.write
Grant Type: Client Credentials
```


postman2


You can also use the access token as header parameter in your rest request.
Please add a new header:
- Key: Authorization
- Value: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsic3czNjAtUkVTVC1BUEkiXSwidXNlcl9uYW1lIjoiYWRtaW5Ac3czNjAub3JnIiwic2NvcGUiOlsic3czNjAucmVhZCIsInN3MzYwLndyaXRlIl0sImV4cCI6MTUxMTgxMTU1NSwiYXV0aG9yaXRpZXMiOlsiUk9MRV9TVzM2MF9VU0VSIl0sImp0aSI6IjI4ZWI0ZTE0LTkzZGYtNDYxZC1iZjhhLTIwYjQ1OGQxOTc1OCIsImNsaWVudF9pZCI6InRydXN0ZWQtc3czNjAtY2xpZW50In0.Cbaqzyq6Yr7oM4Me0SSmeSbjypDZwHwOuJiVlZ4Ee98cdvyhCr7dI3ZV9-zKaiwrcnB9ERdJuoHcTOr1yMK8yvVlRMxwlkSE0QQvIyYUP0KIvR3ZuSTn25rl5inhBzy9usibJ9aQB4MMZOyhvoajuRWL09h0_Dx_gxjJDBMa8MNiNjXzIiF9s5YZbFPGlTaRWoXrTOZAvoUnJXlI_e6q9aMemLa7YGtSERte52LopAfRMpAmgiAJJD_NBijDFyozzhG2wlw6Xdk-B7wEfE7n2TjIhr05KXHYQcNeDhxScFGw_p9jInfm7-gy1q2rmoH_zlMcNedofGvz7uMewvLdew

**As value you need to enter: "Bearer " + ACCESS_TOKEN**

postman3

**Comment by alexbrdn** [comment]
You do plan to stick the endpoints behind nginx, don't you? I think they should be accessible over TLS, not plain HTTP.

**Comment by maxhbr** [comment]
@alexbrdn I think the question whether the service runs behind an https proxy is not handled within sw360portal. This will be the default behaviour in sw360chores.

**Comment by alexbrdn** [comment]
Oh, right

**Comment by maierthomas** [comment]
@alexbrdn I updated my last post.
Of course its working also with HTTPS.

**Comment by maierthomas** [comment]
I need to a cleanup for some files and refactoring here.
But first of all we are able to test the rest features within the sw360 portal.

**Comment by mcjaeger** [comment]
hm, plain install does not work at submitting the curl request already:
```
<html><head><title>Apache Tomcat/7.0.67 - Error report</title><style><!--H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}HR {color : #525D76;}--></style> </head><body><h1>HTTP Status 404 - /authorization/oauth/token</h1><HR size="1" noshade="noshade"><p><b>type</b> Status report</p><p><b>message</b> <u>/authorization/oauth/token</u></p><p><b>description</b> <u>The requested resource is not available.</u></p><HR size="1" noshade="noshade"><h3>Apache Tomcat/7.0.67</h3></body></html>
```

**Comment by maierthomas** [comment]
@adrian-evo @mcjaeger please retest, something was wrong with my last push.
The 404 issue should be resolved now.

**Comment by maierthomas** [comment]
We should add a note in the sw360portal documentation.
The current REST api implementation is in an experimental state and its may exposed by breaking changes.

Changes (e.g. client-id, client-secret...) will be made in further pull request.

**Comment by maierthomas** [comment]
changed readme and added wiki page.
https://github.com/sw360/sw360portal/wiki/REST-API (in processing)

**Pull request merged**